src/extras/primitives/registerPrimitive.js

```diff
@@ -75,6 +69,33 @@ module.exports = function registerPrimitive (name, definition) {
         }
       },
 
+      applyDefaultComponents: {
+        value: function () {
+          var self = this;
+          var defaultData = cloneObject(this.defaultAttributes);
+
+          // Apply default components.
+          Object.keys(defaultData).forEach(function applyDefault (componentName) {
+            var componentData = defaultData[componentName];
+
+            // Set component properties individually to not overwrite user-defined components.
+            if (componentData instanceof Object && Object.keys(componentData).length) {
+              Object.keys(componentData).forEach(function setProperty (propName) {
+                // Check if component property already defined.
+                var definedData = self.getAttribute(componentName);
+                if (definedData && definedData[propName] !== undefined) { return; }
```

Comment:
but couldn't `null` be a valid value for some component properties? probably not with primitives. you know better than I do.